src/components/AutoSuggest/index.js

```diff
@@ -0,0 +1,252 @@
+ */
+const AutoSuggest = ({
+  className,
+  formatResponse = formatFunc,
+  numOptionsDeterminer = numOptionsFunc,
+  onSelect,
+  url = URL,
+}) => {
+  const { get, data } = useFetch();
+  const [show, setShow] = useState(false);
+  const [activeIndex, setActiveIndex] = useState(-1);
+  const wrapperRef = useRef(null);
+  const inputRef = useRef(null);
+
+  const handleKeyDown = useCallback(
+    event => {
+      const numberOfOptions = numOptionsDeterminer(data);
+      if (!show) return;
+
+      switch (event.keyCode) {
+        // Arrow up
+        case 38:
+          event.preventDefault();
+
+          setActiveIndex(state => {
+            const indexOfActive = state - 1;
+            const topReached = indexOfActive === -1;
+
+            return topReached ? numberOfOptions - 1 : indexOfActive;
+          });
+          break;
+
+        // Arrow down
+        case 40:
+          event.preventDefault();
+
+          setActiveIndex(state => {
+            const indexOfActive = state + 1;
+            const endReached = indexOfActive === numberOfOptions;
+
+            return endReached ? 0 : indexOfActive;
+          });
+          break;
+
+        // Enter
+        case 13:
+          if (event.target !== inputRef.current) {
+            inputRef.current.value = event.target.innerText;
+          }
+
+          setActiveIndex(-1);
+          setShow(false);
+          break;
+
+        // Escape
+        case 27:
+          inputRef.current.value = '';
+          setActiveIndex(-1);
+          setShow(false);
+          break;
+
+        // Home
+        // Moves focus to the textbox and places the editing cursor at the beginning of the field.
+        case 36:
+          event.preventDefault();
+
+          inputRef.current.focus();
+          inputRef.current.setSelectionRange(0, 0);
+          break;
+
+        // End
+        // Moves focus to the textbox and places the editing cursor at the end of the field.
+        case 35:
+          event.preventDefault();
+
+          inputRef.current.focus();
+          inputRef.current.setSelectionRange(9999, 9999);
+          break;
+
+        default:
+          setActiveIndex(-1);
+          break;
+      }
+    },
+    [data, numOptionsDeterminer, show]
+  );
+
+  useEffect(() => {
+    if (activeIndex === -1) {
+      inputRef.current.focus();
+    }
+  }, [activeIndex]);
+
+  useEffect(() => {
+    const wrapper = wrapperRef.current;
+
+    wrapper.addEventListener('keydown', handleKeyDown);
+
+    return () => {
+      wrapper.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [handleKeyDown]);
+
+  useEffect(() => {
+    const hasResults = numOptionsDeterminer(data) > 0;
+
+    setShow(hasResults);
+  }, [data, numOptionsDeterminer]);
+
+  const delayedCallback = useCallback(
+    debounce(value => {
+      serviceRequest(value);
+    }, INPUT_DELAY),
+    [serviceRequest]
+  );
+
+  const onChange = useCallback(
+    event => {
+      event.persist();
```

Comment:
why is this needed here?

Reply:
If I don't persist the synthetic event, I cannot get a reference to its `target` prop.